packages/edit-navigation/src/components/menu-editor/index.js

```diff
@@ -6,66 +6,81 @@ import {
 	BlockEditorProvider,
 } from '@wordpress/block-editor';
 import { useViewportMatch } from '@wordpress/compose';
-import { useMemo } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
-import useMenuItems from './use-menu-items';
-import useNavigationBlocks from './use-navigation-blocks';
 import MenuEditorShortcuts from './shortcuts';
 import BlockEditorArea from './block-editor-area';
 import NavigationStructureArea from './navigation-structure-area';
+import useNavigationBlockEditor from './use-navigation-block-editor';
+import useStubPost from './use-stub-post';
+import { useMemo } from '@wordpress/element';
 
 export default function MenuEditor( {
 	menuId,
 	blockEditorSettings,
 	onDeleteMenu,
 } ) {
-	const isLargeViewport = useViewportMatch( 'medium' );
 	const query = useMemo( () => ( { menus: menuId, per_page: -1 } ), [
 		menuId,
 	] );
-	const {
-		menuItems,
-		eventuallySaveMenuItems,
-		createMissingMenuItems,
-	} = useMenuItems( query );
-	const { blocks, setBlocks, menuItemsRef } = useNavigationBlocks(
-		menuItems
-	);
-	const saveMenuItems = () => eventuallySaveMenuItems( blocks, menuItemsRef );
+	const postId = useStubPost( query );
 
 	return (
 		<div className="edit-navigation-menu-editor">
 			<BlockEditorKeyboardShortcuts.Register />
 			<MenuEditorShortcuts.Register />
 
-			<BlockEditorProvider
-				value={ blocks }
-				onInput={ ( updatedBlocks ) => setBlocks( updatedBlocks ) }
-				onChange={ ( updatedBlocks ) => {
-					createMissingMenuItems( updatedBlocks, menuItemsRef );
-					setBlocks( updatedBlocks );
-				} }
-				settings={ {
-					...blockEditorSettings,
-					templateLock: 'all',
-					hasFixedToolbar: true,
-				} }
-			>
-				<BlockEditorKeyboardShortcuts />
-				<MenuEditorShortcuts saveBlocks={ saveMenuItems } />
-				<NavigationStructureArea
-					blocks={ blocks }
-					initialOpen={ isLargeViewport }
-				/>
-				<BlockEditorArea
-					saveBlocks={ eventuallySaveMenuItems }
+			{ postId && (
+				<NavigationBlockEditorProvider
+					query={ query }
+					postId={ postId }
 					menuId={ menuId }
+					blockEditorSettings={ blockEditorSettings }
 					onDeleteMenu={ onDeleteMenu }
 				/>
-			</BlockEditorProvider>
+			) }
 		</div>
 	);
 }
+
+const NavigationBlockEditorProvider = ( {
+	query,
+	menuId,
+	postId,
+	blockEditorSettings,
+	onDeleteMenu,
+} ) => {
+	const isLargeViewport = useViewportMatch( 'medium' );
+	const [
+		blocks,
+		onInput,
+		onChange,
+		saveMenuItems,
+	] = useNavigationBlockEditor( query, postId );
```

Comment:
Instead of having custom `block`/`onInput`/`onChange` handlers, this uses the same ones as the regular post editor.